src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -435,12 +435,15 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                 {
                     const int shiftAmount = helper.ImmValue();
 
-                    if (shiftAmount == 0)
+                    if ((shiftAmount == 0) && (intrin.category == HW_Category_ShiftRightByImmediate))
                     {
-                        // TODO: Use emitIns_Mov instead.
-                        //       We do not use it currently because it will still elide the 'mov'
-                        //       even if 'canSkip' is false. We cannot elide the 'mov' here.
-                        GetEmitter()->emitIns_R_R_R(INS_mov, emitTypeSize(node), targetReg, reg, reg);
+                        // For operations like `Vector128<byte> >> 8`, over-shifting behavior will mask
+                        // the immediate so that it's zero, and the shift will be imported as a GT_HWINTRINSIC.
+                        // If this isn't const-folded away, it is possible to try to emit a right-shift by zero,
+                        // but only if the immediate is known (we won't emit a zero case for the jump table).
+                        // Since we cannot encode a right-shift by 0 on ARM64, emit a mov instead.
+                        assert(op->isContainedIntOrIImmed());
+                        GetEmitter()->emitIns_Mov(INS_mov, emitTypeSize(node), targetReg, reg, /* canSkip */ true);
                     }
```

Comment:
Why wouldn't we emit a case for the jump table?

If the value is unknown and it doesn't throw, then won't we hit the case in the jump table and need handling emitted for it?

Reply:
You're correct that for unknown valid immediates, we will emit the jump table. In the table generation logic, we get the table's bounds from `HWIntrinsicInfo::lookupImmBounds`, which sets the lower bound at 1 for right-shift intrinsics [here](https://github.com/dotnet/runtime/blob/66e45ef152002a2e1cbea939ded6643dfa48307c/src/coreclr/jit/hwintrinsicarm64.cpp#L408). So if `shiftAmount == 0` for a right-shift intrinsic, we should not be generating a jump table.

Reply:
So what are we doing if the value is unknown and `0`? Because we're not throwing for the constant case here.

Reply:
Good question. I tried the following program:
```csharp
public class Test
{
    public byte m;
}

public class Program
{
    static Test t = new Test();
    public static void Main()
    {
        var v = Vector128<Byte>.Zero;
        var result = v >> t.m;
        Console.WriteLine(result);
    }
}
```

The JIT imports the shift as `ShiftLogical`, hence why we don't throw.

Reply:
Hmmm. I wonder if this is the "right" fix.

We basically have:
* Import the intrinsic
* If the shift amount is non-constant, import as `ShiftLogical`
* If the shift amount is constant, then mask, which could become zero

So I think we're maybe missing a check that ensures that if its zero, its still treated as "out of range" and gets imported as `ShiftLogical` instead

Reply:
I think your proposed fix is better. I initially wanted the fallback implementations we've been adding to kick in for cases like this, but the logic for checking immediates and adding range checks/etc. in `Compiler::impHWIntrinsic` runs before we check if the intrinsic is table-driven or needs special handling. I guess in `GenTreeHWIntrinsic::GetHWIntrinsicIdForBinOp`, instead of just checking if the immediate is const, we need to also check if it's out-of-range and use `ShiftLogical` accordingly -- I'll give that a try.